### PR TITLE
Rename jsonCompiler to libsolc

### PR DESCRIPTION
**Pull request by axic**
Part of #2864.

### CODE REVIEW

**Comment by chriseth** [comment]
I think this might break the emscripten deploy step (also solc-js has to be adapted).

**Comment by axic** [comment]
I'd just keep the name as `soljson` for solc-js/solc-bin in this PR. We could consider renaming that part separately to this to minimize dependencies.

**Comment by axic** [comment]
@chfast the executable `solc` and the library `libsolc` are conflicting at link time (i.e. solfuzzer cannot decide which one to link to). Can this be fixed in cmake?

**Comment by chfast** [comment]
If you use the same target names you will have the problem.
There are 2 solutions:

1. Use common pattern of starting library names with capital letter, `libSolc.a`.
2. Use target name "libsolc" and set the output name to "solc" for it.
   ```
   set_target_properties(libsolc PROPERTIES OUTPUT_NAME solc)
   ```

**Comment by axic** [comment]
I am actually using `libsolc` as target name. Can you have a look at the changes?

File: libsolc/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+set(
```
Comment (chfast) [inline libsolc/CMakeLists.txt]: Not needed?

File: libsolc/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+	sources
+)
+
+if (EMSCRIPTEN)
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s EXPORTED_FUNCTIONS='[\"_compileJSON\",\"_license\",\"_version\",\"_compileJSONMulti\",\"_compileJSONCallback\",\"_compileStandard\"]' -s RESERVED_FUNCTION_POINTERS=20")
+	add_executable(libsolc libsolc.cpp)
+else()
+	add_library(libsolc libsolc.cpp)
```
Comment (chfast) [inline libsolc/CMakeLists.txt]: Without changing the `OUTPUT_NAME` you will get a file of name `liblibsolc.a`.

File: libsolc/CMakeLists.txt (hunk 1 above)
Comment (axic) [reply libsolc/CMakeLists.txt]: I was wondering about this. The same applies to `solc/CMakeLists.txt` too I guess.

File: scripts/bytecodecompare/storebytecode.sh [hunk 3]
```diff
@@ -40,7 +40,7 @@ TMPDIR=$(mktemp -d)
 
     if [[ "$SOLC_EMSCRIPTEN" = "On" ]]
     then
-        cp "$REPO_ROOT/build/solc/soljson.js" .
+        cp "$REPO_ROOT/build/libsolc/libsolc.js" .
```
Comment (axic) [inline scripts/bytecodecompare/storebytecode.sh]: `.` needs to be replaced with `soljson.js`

File: libsolc/CMakeLists.txt (hunk 1 above)
Comment (chfast) [inline libsolc/CMakeLists.txt]: You forgot to include this script in main CMakeLists.txt script. CMake does not know what `libsolc` is and it assumed its some system library.

**Review by chfast — CHANGES_REQUESTED** [review]

**Comment by axic** [comment]
The properties thing doesn't work here:
```
  Cannot find source file:

    PROPERTIES

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
```

**Comment by chfast** [comment]
This is a separated command:
```
add_library(libsolc ...)
set_target_properties(libsolc PROPERTIES OUTPUT_NAME solc)
```

File: libsolc/CMakeLists.txt (hunk 1 above)
Comment (chfast) [reply libsolc/CMakeLists.txt]: The `set()` just sets the variable value. It is `sources = ""`. If `sources` are not used anywhere remove it.

**Comment by axic** [comment]
My bad. It seems to work perfectly now, thanks!

File: libsolc/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+if (EMSCRIPTEN)
+	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s EXPORTED_FUNCTIONS='[\"_compileJSON\",\"_license\",\"_version\",\"_compileJSONMulti\",\"_compileJSONCallback\",\"_compileStandard\"]' -s RESERVED_FUNCTION_POINTERS=20")
+	add_executable(soljson libsolc.cpp)
+	target_link_libraries(soljson PRIVATE solidity)
+else()
+	add_library(libsolc SHARED libsolc.cpp)
```
Comment (axic) [inline libsolc/CMakeLists.txt]: The `SHARED` part could be skipped for this PR.

Comment (chfast) [reply libsolc/CMakeLists.txt]: Why do you need it to be a shared library?

Comment (axic) [reply libsolc/CMakeLists.txt]: Ultimate goal of #2864.

Comment (chfast) [reply libsolc/CMakeLists.txt]: So let's keep it here now.

**Review by chfast — APPROVED** [review]

**Comment by axic** [comment]
@chriseth happy to merge?

**Comment by chriseth** [comment]
needs rebase

**Review by chriseth — APPROVED** [review]

**Pull request merged**